cpp/ql/src/semmle/code/cpp/exprs/Call.qll

```diff
@@ -2,12 +2,22 @@ import semmle.code.cpp.exprs.Expr
 import semmle.code.cpp.Function
 private import semmle.code.cpp.dataflow.EscapesTree
 
+private class TCall = @funbindexpr or @callexpr;
+
 /**
  * A C/C++ call.
- *
- * This is the abstract root QL class for all types of calls.
  */
-abstract class Call extends Expr, NameQualifiableElement {
+class Call extends Expr, NameQualifiableElement, TCall {
+  // `@funbindexpr` (which is the dbscheme type for FunctionCall) is a union type that includes
+  // `@routineexpr. This dbscheme type includes accesses to functions that are not necessarily calls to
+  // that function. That's why the charpred for `FunctionCall` requires:
+  // ```
+  // iscall(underlyingElement(this), _)
+  // ```
+  // So for the charpred for `Call` we include the requirement that if this is an instance of
+  // `@funbindexpr` it must be a _call_ to the function.
+  Call() { this instanceof @funbindexpr implies iscall(underlyingElement(this), _) }
```

Comment:
```suggestion
  Call() { this instanceof @callexpr or iscall(underlyingElement(this), _) }
```

I hope this is equivalent.

We try to avoid `implies` because it tends to have slightly worse performance and readability than a more positively-phrased formula.

Reply:
Agreed, this should be equivalent. I've rephrased it as you suggested.